packages/common-common/src/permissions.ts

```diff
@@ -1,63 +1,199 @@
+  // Delete Subtree
+  [Action.DELETE_TOPIC, [Action.DELETE_THREAD]],
+  [Action.DELETE_THREAD, [Action.DELETE_COMMENT]],
+  [Action.DELETE_COMMENT, [Action.DELETE_REACTION]],
+  // Edit Subtree
+  [Action.EDIT_TOPIC, [Action.EDIT_THREAD]],
+  [Action.EDIT_THREAD, [Action.LINK_THREAD_TO_THREAD, Action.LINK_PROPOSAL_TO_THREAD, Action.EDIT_COMMENT]],
+]);
 
-export enum PermissionError {
-  NOT_PERMITTED = 'Action not permitted',
+const DENY_IMPLICT_PERMISSIONS_BY_ACTION = new Map<number, Action[]>([
+  // Chat Subtree
+  [Action.VIEW_CHAT_CHANNELS, [Action.CREATE_CHAT]],
+  // View Subtree
+  [Action.VIEW_REACTIONS, [Action.VIEW_COMMENTS, Action.CREATE_REACTION]],
+  [Action.VIEW_COMMENTS, [Action.VIEW_POLLS, Action.CREATE_COMMENT]],
+  [Action.VIEW_POLLS, [Action.VIEW_THREADS, Action.CREATE_POLL]],
+  [Action.VIEW_THREADS, [Action.VIEW_TOPICS, Action.CREATE_THREAD]],
+  [Action.VIEW_TOPICS, [Action.CREATE_TOPIC]],
+  // Create Subtree
+  [Action.CREATE_REACTION, [Action.VIEW_COMMENTS, Action.DELETE_REACTION]],
+  [Action.CREATE_COMMENT, [Action.VOTE_ON_POLLS, Action.EDIT_COMMENT, Action.DELETE_COMMENT]],
+  [Action.CREATE_POLL, [Action.VIEW_THREADS]],
+  [Action.CREATE_THREAD, [Action.VIEW_TOPICS, Action.EDIT_THREAD, Action.DELETE_THREAD, Action.LINK_PROPOSAL_TO_THREAD, Action.LINK_THREAD_TO_THREAD]],
+  [Action.CREATE_TOPIC, [Action.EDIT_TOPIC, Action.DELETE_TOPIC]],
+  // Voting Subtree
+  [Action.VOTE_ON_POLLS, [Action.CREATE_POLL]],
+  // Delete Subtree
+  [Action.DELETE_REACTION, [Action.DELETE_COMMENT]],
+  [Action.DELETE_COMMENT, [Action.DELETE_THREAD]],
+  [Action.DELETE_THREAD, [Action.DELETE_TOPIC]],
+  // Edit Subtree
+  [Action.EDIT_COMMENT, [Action.EDIT_THREAD]],
+  [Action.EDIT_THREAD, [Action.EDIT_TOPIC]]
+]);
+
+// Recursive function to get all implicit permissions of an action
+const recurseImplicitActions = (action: Action, result_actions: Action[], allowDeny: boolean): Action[] => {
+  let implicitActions;
+  if (allowDeny) {
+    implicitActions = ALLOW_IMPLICIT_PERMISSIONS_BY_ACTION.get(action);
+  } else {
+    implicitActions = DENY_IMPLICT_PERMISSIONS_BY_ACTION.get(action);
+  }
+  // Base Case, if there are no implicit permission leaves, return the action
+  if (!implicitActions) {
+    return [ action ];
+  }
+  // Recursive Case, if there are implicit permission leaves, return the action and the leaves
+  for (let i = 0; i < implicitActions.length; i++) {
+    if (allowDeny) {
+      result_actions = result_actions.concat(recurseImplicitActions(implicitActions[i], ALLOW_IMPLICIT_PERMISSIONS_BY_ACTION.get(implicitActions[i]), allowDeny));
+    } else {
+      result_actions = result_actions.concat(recurseImplicitActions(implicitActions[i], DENY_IMPLICT_PERMISSIONS_BY_ACTION.get(implicitActions[i]), allowDeny));
+    }
+  };
+
+  let uniqueActions = [...new Set(result_actions.concat([action]))];
+  return uniqueActions;
 }
 
+export const getImplicitActionsSet = (action: Action, allowDeny: boolean): Action[] => {
```

Comment:
Is this helper function necessary if we already get the implicit actions on the provided action in the first few lines of `recurseImplicitActions`?

Reply:
I made this just so that we can get abstract out the array parameter